### ISSUE TITLE
Provide mini API for cookie consent plugins to integrate with

### ISSUE THREAD

**Issue by felixarntz** [google/site-kit-wp#2087]
There are a myriad of WordPress plugins that implement a cookie banner or other way of consent in the frontend of a WordPress sites. Some of these plugins also include functionality to integrate with certain plugins so that features that require the user's consent are not added until that consent is granted. Site Kit should expose a simple layer for such plugins to prevent Site Kit's Google service snippets from being rendered on a page if the user has not granted consent.

While theoretically for Analytics it would be sufficient to rely on the existing `googlesitekit_is_tracking_disabled` filter, there should be a broader solution that prevents output of those snippets entirely.

---------------

_Do not alter or remove anything below. The following sections will be managed by moderators only._

## Acceptance criteria

* Filters `googlesitekit_{$slug}_tag_blocked` should be fired before potentially rendering a Google service tag snippet in the frontend for non-AMP, where `$slug` is the service name (`adsense`, `analytics`, `tagmanager`). The filterable parameter should be a boolean (default `false`).
    * All non-AMP tag snippet logic should only be run if the filter value is `false`.
* Filters `googlesitekit_{$slug}_tag_amp_blocked` should be fired before potentially rendering a Google service tag snippet in the frontend for AMP, where `$slug` is the service name (`adsense`, `analytics`, `tagmanager`). The filterable parameter should be a boolean (default `false`).
    * All AMP tag snippet logic should only be run if the filter value is `false`.
* [Since AMP includes a standardized layer for blocking snippets based on a consent](https://amp.dev/documentation/components/amp-consent/#basic-blocking-behaviors), more fine-grained support on the client-side is available, and it should be exposed via a server-side `googlesitekit_{$slug}_tag_amp_block_on_consent` filter, where `$slug` is the service name (`adsense`, `analytics`, `tagmanager`). The filterable parameter should be `false` by default, but can also be `true` or one of the strings `_till_responded`, `_till_accepted`, or `_auto_reject`.
    * If the filter value is `true` or one of the three strings, the AMP component tags should receive a `data-block-on-consent` attribute (if one of the three strings, with that value).
* For non-AMP, there is no client-side standardization, but there should still be a server-side `googlesitekit_{$slug}_tag_block_on_consent` filter, where `$slug` is the service name. The filterable parameter should be `false` by default but can also be `true`.
    * If the filter value is `true`, all `<script>` tags rendered in the frontend as part of the respective module would receive a `type="text/plain"` attribute so that the JavaScript doesn't evaluate. They should also receive a `data-block-on-consent` attribute. The 3P plugin could then implement its own client-side logic to change or remove the attribute so that the browser starts executing it.

## Implementation Brief

* Implement a trait `Module_With_Blockable_Tags_Trait`:
    * Add protected method `is_tag_blocked()`:
        * If AMP, call `return apply_filters( 'googlesitekit_{$this->slug}_tag_amp_blocked', false );`.
        * Otherwise, call `return apply_filters( 'googlesitekit_{$this->slug}_tag_blocked', false );`.
    * Add protected method `get_tag_amp_block_on_consent_attribute()`:
        * Call `$block_on_consent = apply_filters( 'googlesitekit_{$this->slug}_tag_amp_block_on_consent', false );`.
        * If `true`, return ` data-block-on-consent`.
        * If one of the three strings, return ` data-block-on-consent="{string}"`.
        * Otherwise, return an empty string.
    * Add protected method `get_tag_block_on_consent_attribute()`:
        * Call `$block_on_consent = apply_filters( 'googlesitekit_{$this->slug}_tag_block_on_consent', false );`.
        * If `true`, return ` type="text/plain" data-block-on-consent`.
        * Otherwise, return an empty string.
    * Note that this trait is related to the proposal in #475 which should not be done at this point, but will be fully interoperable.
* Make the `AdSense`, `Analytics`, and `Tag_Manager` classes use the new trait:
    * At the beginning of the `template_redirect` hook, call `is_tag_blocked()` and bail early if `true`.
    * In all methods that render AMP components, call `get_tag_amp_block_on_consent_attribute()` (e.g. `<amp-analytics<?php echo $this->get_tag_amp_block_on_consent_attribute(); ?>>`).
    * In all methods that render frontend scripts, call `get_tag_block_on_consent_attribute()` (e.g. `<script<?php echo $this->get_tag_block_on_consent_attribute(); ?>>`).
* Expand existing PHPUnit coverage to include assertions for the new filters for each of the three classes.

## QA Brief

This issue affects both AMP and non-AMP tag output for the Analytics, Tag Manager, and AdSense modules.

The changes introduced here don't affect the setup at all so they can all be set up ahead of time.

Since this is only support added for 3rd parties to leverage, no change is visible without adding additional code to enable it.

Snippets to use for testing tag blocking (prevents tags from being output entirely):
```php
$slug = 'analytics'; // Also: adsense | tagmanager
// Block web and AMP tags
add_filter( "googlesitekit_{$slug}_tag_blocked", '__return_true' );
// Block AMP tag only (AMP plugin must be active)
add_filter( "googlesitekit_{$slug}_tag_amp_blocked", '__return_true' );
```
Snippets to use for testing "block on consent" behavior: tag HTML will be rendered on the page but loading is deferred until user consents (must be implemented by 3p)
```php
// AMP tag consent blocking: amp tags will get a `data-block-on-consent` attribute using below filter.
add_filter( "googlesitekit_{$slug}_tag_amp_block_on_consent", '__return_true' );
// AMP tag consent blocking: amp tags will get a `data-block-on-consent="one of values below only"` attribute using below filter.
$block_on_consent = 'one of these ->' // _till_responded | _till_accepted | _auto_reject (if any other truthy value is passed, the value will be omitted leaving only the boolean data-block-on-consent attribute)
add_filter( "googlesitekit_{$slug}_tag_amp_block_on_consent", fn() => $block_on_consent );
// --- Block on consent for non-AMP
// On/off only. When enabled, the primary script for the tag will get a `data-block-on-consent` (no value) and a `type="text/plain"` attributes.
add_filter( "googlesitekit_{$slug}_tag_block_on_consent", '__return_true' );
```

**Test**
* AdSense (AMP & non-AMP)
* Analytics (AMP & non-AMP)
* Tag Manager (AMP & non-AMP)

## Changelog entry

* Introduce several filters to allow blocking rendering or execution of Google service scripts in the frontend, allowing integration with e.g. 3P cookie consent plugins.

**Comment by aaemnnosttv** [comment]
Looks good for the most part @felixarntz. A few things seem to need revision:

> Implement a trait `Module_With_Tags`

All of our current `Module_With_` names apply to interfaces. Should we not also add a `Module_With_Tags` interface here in addition to a `Module_With_Tags_Trait`?

> * If AMP, call `return apply_filters( 'googlesitekit_{$this->slug}_tag_blocked', false );`.
        * Otherwise, call `return apply_filters( 'googlesitekit_{$this->slug}_tag_amp_blocked', false );`.

It looks like you have the filter names backwards for these conditions.

Also, just curious how this works with the cookie consent since this affects the resulting HTML. When a user clicks the consent banner, the plugin would set a cookie and trigger a page refresh in order for the "consented" HTML to be loaded? I'm guessing AMP has another mechanism which doesn't require the reload but for non-AMP?

**Comment by felixarntz** [comment]
@aaemnnosttv 
> All of our current `Module_With_` names apply to interfaces. Should we not also add a `Module_With_Tags` interface here in addition to a `Module_With_Tags_Trait`?

As part of #475 yes, but the trait so far will only have protected methods, hence no need for an interface I'd say.

> Also, just curious how this works with the cookie consent since this affects the resulting HTML. When a user clicks the consent banner, the plugin would set a cookie and trigger a page refresh in order for the "consented" HTML to be loaded? I'm guessing AMP has another mechanism which doesn't require the reload but for non-AMP?

Yeah this API would be limited to server-side integration only which indeed requires a page refresh. For AMP, the `data-block-on-consent` attribute can be used to automatically handle this client-side, so that is basically a better alternative approach to using `googlesitekit_{$this->slug}_tag_amp_blocked`, which is also server-side. Due to the lack of standardization of client-side functionality here outside of AMP, I don't think it makes sense to offer a solution here because it would need to cater for a myriad of different ways that the consent plugins could behave.

**Comment by aaemnnosttv** [comment]
> As part of #475 yes, but the trait so far will only have protected methods, hence no need for an interface I'd say.

Do you mean then that in #475 `Module_With_Tags` would become an interface ? If we don't need an interface right now that's fine, and makes sense if we're only introducing protected methods, but I think we should stick to the `Module_With_X` interface and `Module_With_X_Trait` naming convention.

> Due to the lack of standardization of client-side functionality here outside of AMP, I don't think it makes sense to offer a solution here because it would need to cater for a myriad of different ways that the consent plugins could behave.

My only thought here would be that rather than blocking the output of the tag entirely, non-AMP could use a similar strategy and render the tag with a `type="text/plain"` attribute or [`template`](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/template) tag. That way, the HTML/js content would still be there and would better support full-page caching. The client would need to implement their own "activation" of the script based on consent (for non-AMP), but would still allow for a better client-side experience that didn't require a full-page refresh. What do you think?

**Comment by felixarntz** [comment]
@aaemnnosttv 
> Do you mean then that in #475 `Module_With_Tags` would become an interface ? If we don't need an interface right now that's fine, and makes sense if we're only introducing protected methods, but I think we should stick to the `Module_With_X` interface and `Module_With_X_Trait` naming convention.

Ah I see, I missed the naming point, updated.

> My only thought here would be that rather than blocking the output of the tag entirely, non-AMP could use a similar strategy and render the tag with a `type="text/plain"` attribute or [`template`](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/template) tag. That way, the HTML/js content would still be there and would better support full-page caching. The client would need to implement their own "activation" of the script based on consent (for non-AMP), but would still allow for a better client-side experience that didn't require a full-page refresh. What do you think?

I like that idea. Updated the ACs and IB accordingly, adding a `googlesitekit_{$slug}_tag_block_on_consent` filter. I think the name makes sense for consistency even if the attribute is technically not called that (like it is in AMP - but even there that's more of an implementation detail from the server-side).

**Comment by aaemnnosttv** [comment]
Awesome! What do you think about adding a boolean `data-block-on-consent` attribute when blocked as well? It would help identify which scripts have been blocked rather than looking at the `type` alone.

> * If the filter value is `true`, all `<script>` tags rendered in the frontend as part of the respective module would receive a `type="text/plain"` attribute so that the JavaScript doesn't evaluate. The 3P plugin could then implement its own client-side logic to change or remove the attribute so that the browser starts executing it.

I thought it would be that easy, but simply changing the `type` attribute after the tag has been evaluated won't cause it to be executed. You actually need to do something like this (where `sourceEl` is the disabled script):

```js
function loadScript( sourceEl ) {
    const targetEl = document.createElement( 'script' );
    targetEl.innerHTML = sourceEl.innerHTML;
    sourceEl.parentElement.appendChild( targetEl ); // loads the script once added
    // optionally remove sourceEl
}
```

I don't think that's a dealbreaker or anything here but just wanted to point out that it takes a little more than just modifying the disabled `script` tag in place.

Otherwise, ✅

**Comment by aaemnnosttv** [comment]
IB ✅

**Comment by aaemnnosttv** [comment]
@felixarntz this is coming along well and is mostly done, but I'm running into a bit of an unexpected complication when it comes to adding the attribute(s) to the tag. There isn't really a problem with outputting the HTML since that's very straightforward, but it's a bit problematic from a security/output-escaping perspective since we'd be including a "black box" of html attributes in the tag. WP of course has `esc_attr` but this is for attribute values and there isn't a good native option for handling a combination of attribute names and values other than `wp_kses` but that seems a bit heavy-handed and it feels wrong to add a `phpcs:ignore` for this kind of thing.

I propose reworking this a bit to use PHP's [DOM API](https://www.php.net/manual/en/book.dom.php) which is better suited to composing bits of HTML.

E.g. The AdSense method for outputting an AMP tag can be updated accordingly:

```php
	protected function print_amp_auto_ads( $client_id ) {
		if ( $this->adsense_tag_printed ) {
			return;
		}

		$this->adsense_tag_printed = true;

		$dom = new DOMDocument();
		$tag = $dom->createElement( 'amp-auto-ads' );
		$tag->setAttribute( 'type', 'adsense' );
		$tag->setAttribute( 'data-ad-client', $client_id );

		$this->apply_amp_block_on_consent( $tag );

		$dom->appendChild( $tag );
		$dom->saveHTMLFile( 'php://stdout' );
	}
```

Note the `apply_amp_block_on_consent` method (which doesn't exist) but could be implemented by the `Module_With_Tags_Trait` instead of `get_tag_amp_block_on_consent_attribute`.

```php
	protected function apply_amp_block_on_consent( DOMElement $tag ) {
		$block_on_consent = apply_filters( "googlesitekit_{$this->slug}_tag_amp_block_on_consent", false );

		if ( in_array( $block_on_consent, $this->get_allowed_block_on_consent_values(), true ) ) {
			$tag->setAttribute( 'data-block-on-consent', $block_on_consent );
		}

		if ( filter_var( $block_on_consent, FILTER_VALIDATE_BOOLEAN ) ) {
			$tag->setAttributeNode( new DOMAttr( 'data-block-on-consent' ) );
		}
	}
```

This way, the method has full control over the tag to apply what it needs using a native API and we don't need to worry about invalid HTML being generated or escaping attribute values since DOM handles this for us. I realize that wouldn't happen as originally defined since we're in full control of that content, so just thinking a little bigger picture here since these methods could potentially be extended by a module.

So my suggestion here is that we would update the IB to use this approach instead (no changes needed to ACs) with all of the output and logic remaining the same, just implemented a bit differently.

```diff
- Module_With_Tags::get_tag_block_on_consent_attribute() : string
+ Module_With_Tags::apply_block_on_consent( DOMElement ) : void
- Module_With_Tags::get_tag_amp_block_on_consent_attribute() : string
+ Module_With_Tags::apply_amp_block_on_consent( DOMElement ) : void
```

What do you think?

**Comment by felixarntz** [comment]
@aaemnnosttv Hmm I'm not sure using the DOM API is worth it here. As you're saying, the way we deal with these attributes here is safe because we're in control and can use `esc_attr` where needed. I'm wary of using the DOM API this way mostly because it's a very different pattern from generating HTML like we do / WordPress does typically (just using strings or echo it), and as far as I remember from previous work there are some minor quirks with that API - plus it reads a bit too verbose for my taste (`$dom->saveHTMLFile( 'php://stdout' )` is also strange).

**Comment by felixarntz** [comment]
@aaemnnosttv Revisiting #475, based on that let's slightly modify the name of `Module_With_Tags_Trait` here as it will conflict with what we should be doing in the other issue. Let's introduce `Module_With_Blockable_Tags_Trait` here, just so that the name doesn't conflict with the "real" `Module_With_Tags_Trait` described in #475.

Once that other issue is in place, we won't need `Module_With_Blockable_Tags_Trait` anymore since it's handled in a different way, which is fine. We should just avoid introducing a trait which we then entirely change. For the work here, this really only means renaming the trait.

**Comment by aaemnnosttv** [comment]
> I'm wary of using the DOM API this way mostly because it's a very different pattern from generating HTML like we do / WordPress does typically (just using strings or echo it), and as far as I remember from previous work there are some minor quirks with that API - plus it reads a bit too verbose for my taste (`$dom->saveHTMLFile( 'php://stdout' )` is also strange)

Hmm, okay. I'm not sure about any quirks but it was just an idea for a potential cleaner approach. As for the output, you can also do `echo $dom->saveHTML()` – the former was more of an experiment which also works 😄 

I'll stick with the original approach going forward.

> Let's introduce `Module_With_Blockable_Tags_Trait` here, just so that the name doesn't conflict with the "real" `Module_With_Tags_Trait` described in #475.

Sounds good 👍

**Comment by tofumatt** [comment]
Chatted to @aaemnnosttv and the QA Brief here changed: turns out now the AMP and non-AMP versions are excluded separately (unlike the QA brief's code that suggests both are disabled by the "Web" filter).

After testing it that way: QA ✅

**Comment by Peebutt1986** [comment]
> There are a myriad of WordPress plugins that implement a cookie banner or other way of consent in the frontend of a WordPress sites. Some of these plugins also include functionality to integrate with certain plugins so that features that require the user's consent are not added until that consent is granted. Site Kit should expose a simple layer for such plugins to prevent Site Kit's Google service snippets from being rendered on a page if the user has not granted consent.
> 
> While theoretically for Analytics it would be sufficient to rely on the existing `googlesitekit_is_tracking_disabled` filter, there should be a broader solution that prevents output of those snippets entirely.
> 
> _Do not alter or remove anything below. The following sections will be managed by moderators only._
> 
> ## Acceptance criteria
> * Filters `googlesitekit_{$slug}_tag_blocked` should be fired before potentially rendering a Google service tag snippet in the frontend for non-AMP, where `$slug` is the service name (`adsense`, `analytics`, `tagmanager`). The filterable parameter should be a boolean (default `false`).
>   
>   * All non-AMP tag snippet logic should only be run if the filter value is `false`.
> * Filters `googlesitekit_{$slug}_tag_amp_blocked` should be fired before potentially rendering a Google service tag snippet in the frontend for AMP, where `$slug` is the service name (`adsense`, `analytics`, `tagmanager`). The filterable parameter should be a boolean (default `false`).
>   
>   * All AMP tag snippet logic should only be run if the filter value is `false`.
> * [Since AMP includes a standardized layer for blocking snippets based on a consent](https://amp.dev/documentation/components/amp-consent/#basic-blocking-behaviors), more fine-grained support on the client-side is available, and it should be exposed via a server-side `googlesitekit_{$slug}_tag_amp_block_on_consent` filter, where `$slug` is the service name (`adsense`, `analytics`, `tagmanager`). The filterable parameter should be `false` by default, but can also be `true` or one of the strings `_till_responded`, `_till_accepted`, or `_auto_reject`.
>   
>   * If the filter value is `true` or one of the three strings, the AMP component tags should receive a `data-block-on-consent` attribute (if one of the three strings, with that value).
> * For non-AMP, there is no client-side standardization, but there should still be a server-side `googlesitekit_{$slug}_tag_block_on_consent` filter, where `$slug` is the service name. The filterable parameter should be `false` by default but can also be `true`.
>   
>   * If the filter value is `true`, all `<script>` tags rendered in the frontend as part of the respective module would receive a `type="text/plain"` attribute so that the JavaScript doesn't evaluate. They should also receive a `data-block-on-consent` attribute. The 3P plugin could then implement its own client-side logic to change or remove the attribute so that the browser starts executing it.
> 
> ## Implementation Brief
> * Implement a trait `Module_With_Blockable_Tags_Trait`:
>   
>   * Add protected method `is_tag_blocked()`:
>     
>     * If AMP, call `return apply_filters( 'googlesitekit_{$this->slug}_tag_amp_blocked', false );`.
>     * Otherwise, call `return apply_filters( 'googlesitekit_{$this->slug}_tag_blocked', false );`.
>   * Add protected method `get_tag_amp_block_on_consent_attribute()`:
>     
>     * Call `$block_on_consent = apply_filters( 'googlesitekit_{$this->slug}_tag_amp_block_on_consent', false );`.
>     * If `true`, return ` data-block-on-consent`.
>     * If one of the three strings, return ` data-block-on-consent="{string}"`.
>     * Otherwise, return an empty string.
>   * Add protected method `get_tag_block_on_consent_attribute()`:
>     
>     * Call `$block_on_consent = apply_filters( 'googlesitekit_{$this->slug}_tag_block_on_consent', false );`.
>     * If `true`, return ` type="text/plain" data-block-on-consent`.
>     * Otherwise, return an empty string.
>   * Note that this trait is related to the proposal in [Refactor tag implementations with polymorphism #475](https://github.com/google/site-kit-wp/issues/475) which should not be done at this point, but will be fully interoperable.
> * Make the `AdSense`, `Analytics`, and `Tag_Manager` classes use the new trait:
>   
>   * At the beginning of the `template_redirect` hook, call `is_tag_blocked()` and bail early if `true`.
>   * In all methods that render AMP components, call `get_tag_amp_block_on_consent_attribute()` (e.g. `<amp-analytics<?php echo $this->get_tag_amp_block_on_consent_attribute(); ?>>`).
>   * In all methods that render frontend scripts, call `get_tag_block_on_consent_attribute()` (e.g. `<script<?php echo $this->get_tag_block_on_consent_attribute(); ?>>`).
> * Expand existing PHPUnit coverage to include assertions for the new filters for each of the three classes.
> 
> ## QA Brief
> This issue affects both AMP and non-AMP tag output for the Analytics, Tag Manager, and AdSense modules.
> 
> The changes introduced here don't affect the setup at all so they can all be set up ahead of time.
> 
> Since this is only support added for 3rd parties to leverage, no change is visible without adding additional code to enable it.
> 
> Snippets to use for testing tag blocking (prevents tags from being output entirely):
> 
> ```
> $slug = 'analytics'; // Also: adsense | tagmanager
> // Block web and AMP tags
> add_filter( "googlesitekit_{$slug}_tag_blocked", '__return_true' );
> // Block AMP tag only (AMP plugin must be active)
> add_filter( "googlesitekit_{$slug}_tag_amp_blocked", '__return_true' );
> ```
> 
> Snippets to use for testing "block on consent" behavior: tag HTML will be rendered on the page but loading is deferred until user consents (must be implemented by 3p)
> 
> ```
> // AMP tag consent blocking: amp tags will get a `data-block-on-consent` attribute using below filter.
> add_filter( "googlesitekit_{$slug}_tag_amp_block_on_consent", '__return_true' );
> // AMP tag consent blocking: amp tags will get a `data-block-on-consent="one of values below only"` attribute using below filter.
> $block_on_consent = 'one of these ->' // _till_responded | _till_accepted | _auto_reject (if any other truthy value is passed, the value will be omitted leaving only the boolean data-block-on-consent attribute)
> add_filter( "googlesitekit_{$slug}_tag_amp_block_on_consent", fn() => $block_on_consent );
> // --- Block on consent for non-AMP
> // On/off only. When enabled, the primary script for the tag will get a `data-block-on-consent` (no value) and a `type="text/plain"` attributes.
> add_filter( "googlesitekit_{$slug}_tag_block_on_consent", '__return_true' );
> ```
> 
> **Test**
> 
> * AdSense (AMP & non-AMP)
> * Analytics (AMP & non-AMP)
> * Tag Manager (AMP & non-AMP)
> 
> ## Changelog entry
> * Introduce several filters to allow blocking rendering or execution of Google service scripts in the frontend, allowing integration with e.g. 3P cookie consent plugins.

**Comment by felixarntz** [comment]
@fabianpimminger I'm not sure there's a reasonable solution to this problem. The scripts are currently all discoverable via `script[data-block-on-consent]`. I understand your concern, there's not a clear way to differentiate what they are for, but we cannot even really put identifiers on them reliably, _especially_ in the Tag Manager case: Tag Manager can be used as a container for various other services, so technically the Tag Manager script could end up loading something related to Analytics, something related to AdSense, etc. So it would be challenging to know which type of consent this script needs.

Analytics and AdSense would be more straightforward, but I'm still unsure about whether this would provide any additional value. If a cookie integration wants to specifically support Site Kit, they could as well differentiate based on the specific script URL, which is different per service.

**Comment by fabianpimminger** [comment]
@felixarntz thanks for the super fast reply. Analytics and Adsense would fall under two different consent categories under gdpr (statistics/analytics and marketing). As far as I understand this regulation, users should be able to consent to (or deny) them individually. So any solution has to load the two services separately. 

I am currently working on an integration for a Consent Plugin and the only solution would be to require a reload (which is a very bad solution) or doing some kind of pattern matching (for example "is previous node a comment and does it have "Google AdSense snippet added by Site Kit" as the value) which is also bad. 

Regarding Tag Manager: You're right. All services loaded via GTM should be out of scope for this integration and should be handled seperately. But regulations are also getting tougher and I think it would also be future proof to make this available for gtm.

**Comment by rlankhorst** [comment]
A more generic approach would be to integrate with https://wordpress.org/plugins/wp-consent-api/, which is a proposal for a universal consent api for WordPress.
